File: ios/BleClient.swift
import Foundation
import CoreBluetooth

class BleClient: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {

    private var centralManager: CBCentralManager!
    private var peripheral: CBPeripheral?

    // Protocol Constants
    static let FFF0_SERVICE_UUID = CBUUID(string: "FFF0")
    static let FFF1_CHAR_UUID = CBUUID(string: "FFF1") // Write
    static let FFF2_CHAR_UUID = CBUUID(string: "FFF2") // Notify

    static let FFE0_SERVICE_UUID = CBUUID(string: "FFE0")
    static let FFE1_CHAR_UUID = CBUUID(string: "FFE1") // Write
    static let FFE2_CHAR_UUID = CBUUID(string: "FFE2") // Notify

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    func connect(device: Device) {
        // Logic to find and connect to the peripheral
    }

    func disconnect() {
        if let peripheral = peripheral {
            centralManager.cancelPeripheralConnection(peripheral)
        }
    }

    // CBCentralManagerDelegate Methods
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        // Handle Bluetooth state updates
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        // Handle device discovery
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        // Handle successful connection
        self.peripheral = peripheral
        self.peripheral?.delegate = self
        self.peripheral?.discoverServices(nil)
    }

    // CBPeripheralDelegate Methods
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        // Handle service discovery
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        // Handle characteristic discovery
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        // Handle notifications
    }

    func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        // Handle write confirmations
    }
    
    // Placeholder for command sending logic
    func sendCommand(command: Data) {
        // Command queue and write logic will be implemented here
    }
}

File: android/src/main/java/com/memotalk/sdk/BleClient.kt
package com.memotalk.sdk

import android.bluetooth.* 
import android.content.Context
import android.os.Handler
import android.os.Looper
import java.util.* 

class BleClient(private val context: Context, private val handler: Handler = Handler(Looper.getMainLooper())) {

    private val bluetoothAdapter: BluetoothAdapter? by lazy { 
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        bluetoothManager.adapter
    }

    private var gatt: BluetoothGatt? = null

    // Protocol Constants
    companion object {
        val FFF0_SERVICE_UUID: UUID = UUID.fromString("0000fff0-0000-1000-8000-00805f9b34fb")
        val FFF1_CHAR_UUID: UUID = UUID.fromString("0000fff1-0000-1000-8000-00805f9b34fb") // Write
        val FFF2_CHAR_UUID: UUID = UUID.fromString("0000fff2-0000-1000-8000-00805f9b34fb") // Notify

        val FFE0_SERVICE_UUID: UUID = UUID.fromString("0000ffe0-0000-1000-8000-00805f9b34fb")
        val FFE1_CHAR_UUID: UUID = UUID.fromString("0000ffe1-0000-1000-8000-00805f9b34fb") // Write
        val FFE2_CHAR_UUID: UUID = UUID.fromString("0000ffe2-0000-1000-8000-00805f9b34fb") // Notify
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt?, status: Int, newState: Int) {
            super.onConnectionStateChange(gatt, status, newState)
            // Handle connection state changes
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt?, status: Int) {
            super.onServicesDiscovered(gatt, status)
            // Handle service discovery
        }

        override fun onCharacteristicWrite(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?, status: Int) {
            super.onCharacteristicWrite(gatt, characteristic, status)
            // Handle write confirmations
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?) {
            super.onCharacteristicChanged(gatt, characteristic)
            // Handle notifications
        }
    }

    fun connect(device: Device) {
        if (bluetoothAdapter == null) {
            // Handle Bluetooth not supported
            return
        }
        gatt = device.nativeDevice.connectGatt(context, false, gattCallback)
    }

    fun disconnect() {
        gatt?.disconnect()
    }

    fun close() {
        gatt?.close()
        gatt = null
    }

    // Placeholder for command sending logic
    fun sendCommand(command: ByteArray) {
        // Command queue and write logic will be implemented here
    }
}
